pkg/provisioner/ironic/update_opts.go

```diff
@@ -0,0 +1,145 @@
+package ironic
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/go-logr/logr"
+
+	"github.com/gophercloud/gophercloud/openstack/baremetal/v1/nodes"
+)
+
+type optionsSection string
+
+const (
+	propertiesOpts   optionsSection = "properties"
+	instanceInfoOpts optionsSection = "instance_info"
+)
+
+type optionsData map[string]interface{}
+
+func optionValueEqual(current, value interface{}) bool {
+	switch curVal := current.(type) {
+	case string:
+		if newStr, ok := value.(string); ok {
+			return curVal == newStr
+		}
+	case bool:
+		if newBool, ok := value.(bool); ok {
+			return curVal == newBool
+		}
+	case int:
+		if newInt, ok := value.(int); ok {
+			return curVal == newInt
+		}
+	case []interface{}:
+		// newType could reasonably be either []interface{} or e.g. []string,
+		// so we muct use reflection.
+		newType := reflect.TypeOf(value)
+		switch newType.Kind() {
+		case reflect.Slice, reflect.Array:
+		default:
+			return false
+		}
+		newList := reflect.ValueOf(value)
+		if newList.Len() != len(curVal) {
+			return false
+		}
+		for i, v := range curVal {
+			if !optionValueEqual(newList.Index(i).Interface(), v) {
+				return false
+			}
+		}
+		return true
+	case map[string]interface{}:
+		// newType could reasonably be either map[string]interface{} or
+		// e.g. map[string]string, so we muct use reflection.
+		newType := reflect.TypeOf(value)
+		if newType.Kind() != reflect.Map ||
+			newType.Key().Kind() != reflect.String {
+			return false
+		}
+		newMap := reflect.ValueOf(value)
+		if newMap.Len() != len(curVal) {
+			return false
+		}
+		for k, v := range curVal {
+			newV := newMap.MapIndex(reflect.ValueOf(k))
+			if !(newV.IsValid() && optionValueEqual(newV.Interface(), v)) {
+				return false
+			}
+		}
+		return true
+	}
+	return false
+}
+
+func getUpdateOperation(name string, currentData map[string]interface{}, desiredValue interface{}, path string, log logr.Logger) *nodes.UpdateOperation {
+	current, present := currentData[name]
+	if desiredValue != nil {
+		if !(present && optionValueEqual(current, desiredValue)) {
+			if log != nil {
+				if present {
+					log.Info("updating option data", "option", name, "value", desiredValue)
+				} else {
+					log.Info("adding option data", "option", name, "value", desiredValue)
+				}
+			}
+			return &nodes.UpdateOperation{
+				Op:    nodes.AddOp, // Add also does replace
+				Path:  path,
+				Value: desiredValue,
+			}
+		}
+	} else {
+		if present {
+			if log != nil {
+				log.Info("removing option data", "option", name)
+			}
+			return &nodes.UpdateOperation{
+				Op:   nodes.RemoveOp,
```

Comment:
Using `nil` to indicate a removal operation can make more difficult to understand the intentions (in the points where the method is used). A dedicated function will be easier to read and less error-prone

Reply:
That makes sense in the cases where we are unconditionally clearing a field, but I [tried this out](https://github.com/zaneb/baremetal-operator/blob/ironic-update-options-fluent/pkg/provisioner/ironic/update_opts.go) and it turns out that there are some cases where we are e.g. getting the value to pass from a pointer that may be nil. In those cases we don't want to have to look at the value before deciding what to call, we just want to pass the value we've been given and have ironic make it right.
Also the code was way too complicated for the amount of lifting it is doing, although that might just have been me going overboard :laughing:

Reply:
Thanks, I understood your point but still remain skeptical in flattening a well defined semantic (add/replace/remove) as the opportunities of unwanted mistakes could increase. The added value of skipping updates when not required still anyhow remains high, so we can eventually re-evaluate this specific point later into another PR  if required.